src/main/scala/viper/gobra/translator/encodings/maps/MathematicalMapEncoding.scala

```diff
@@ -0,0 +1,123 @@
+      for {
+        isCompKey <- checkKeyComparability(idx)(ctx)
+        _ <- assert(isCompKey, comparabilityErrorT) // key must be comparable
+        stmt <- ctx.typeEncoding.assignment(ctx)(in.Assignee(base), in.GhostCollectionUpdate(base, idx, rhs)(src.info), src)
+      } yield stmt
+  }
+
+  /**
+    * Encodes expressions as values that do not occupy some identifiable location in memory.
+    *
+    * To avoid conflicts with other encodings, a leaf encoding for type T should be defined at:
+    * (1) exclusive operations on T, which includes literals and default values
+    *
+    * Most cases are a one-to-one mapping to Viper's sequence operations.
+    * R[ dflt(mmap[K]V°) ] -> EmptyMap([K], [V])
+    * R[ mmapLit(E) ] -> if E.isEmpty then EmptyMap else ExplicitMap([E]) (also checks that the values of keys are all
+    * *   distinct and throws an error if not)
+    * R[ (e: mmap[K]V)[idx] ] -> MapLookup([e], [idx])
+    * R[ (e: mmap[K]V)[idx = val] ] -> MapUpdate([e], [idx], [val])
+    * R[ len(e: mmap[K]V) ] -> MapCardinality([e])
+    * R[ keySet(e: mmap[K]V) ] -> MapDomain([e])
+    * R[ valueSet(e: mmap[K]V) ] -> MapRange([e])
+    */
+  override def expr(ctx: Context): in.Expr ==> CodeWriter[vpr.Exp] = {
+    def goE(x: in.Expr): CodeWriter[vpr.Exp] = ctx.expr.translate(x)(ctx)
+    def goT(t: in.Type): vpr.Type = ctx.typeEncoding.typ(ctx)(t)
+
+    default(super.expr(ctx)){
+      case (e: in.DfltVal) :: ctx.MathematicalMap(k, v) / Exclusive =>
+        unit(withSrc(vpr.EmptyMap(ctx.typeEncoding.typ(ctx)(k), ctx.typeEncoding.typ(ctx)(v)), e))
+
+      case (lit: in.MathMapLit) :: ctx.MathematicalMap(keyT, valT) / Exclusive =>
+        val (pos, info, errT) = lit.vprMeta
+        val vprKeyT = goT(keyT)
+        val vprValT = goT(valT)
+        for {
+          mapletList <- sequence(lit.entries.toVector.map {
+            case (key, value) => for { k <- goE(key); v <- goE(value) } yield vpr.Maplet(k, v)(pos, info, errT)
+          })
+          mmapVal <- if (mapletList.nonEmpty) {
+            // silver assumes that the argument to ExplicitMap is not empty
+            val keySeq = mapletList map (_.key)
+            // checks whether all keys are distinct
+            val checkAllDiffKeys = vpr.EqCmp(
+              vpr.AnySetCardinality(vpr.ExplicitSet(keySeq)(pos, info, errT))(pos, info, errT),
+              vpr.SeqLength(vpr.ExplicitSeq(keySeq)(pos, info, errT))(pos, info, errT)
+            )(pos, info, errT)
+            assert(checkAllDiffKeys, vpr.ExplicitMap(mapletList)(pos, info, errT), repeatedKeyErrorT)(ctx)
+          } else {
+            unit(vpr.EmptyMap(vprKeyT, vprValT)(pos, info, errT))
+          }
+        } yield mmapVal
+
+      case n@ in.IndexedExp(e :: ctx.MathematicalMap(_, _), idx) =>
+        val (pos, info, errT) = n.vprMeta
+        for {
+          isCompKey <- checkKeyComparability(idx)(ctx)
```

Comment:
You are inconsistent with these checks. 

First, I still do not think that keys have to be comparable for mathematical maps. Even if we compare functions, there is always an identity at the verification level. For instance, sets also do not require that the elemnts are comparable. Otherwise, we are unnecessarily restricted in what we can express in Gobra. For instance, we could not have a map from slices to some ghost state, e.g. the number of occurrences of a certain element. 

Second, if you check that keys are comparabile then these checks have to be consistent. You also need to check that the key used in an update expression is comparable and you have to check that the keys used in a literal are comparable. Both of these are missing. (But again, I would just remove the check here).